src/x509_parser.rs

```diff
@@ -521,6 +551,84 @@ pub fn parse_algorithm_identifier(i: &[u8]) -> X509Result<AlgorithmIdentifier> {
     })(i)
 }
 
+fn attributes_sequence_to_map<'a>(
+    i: &'a [u8],
+    v: Vec<X509CriAttribute<'a>>,
+) -> X509Result<'a, HashMap<Oid<'a>, X509CriAttribute<'a>>> {
+    let mut attributes = HashMap::new();
+    for attr in v.into_iter() {
+        if attributes.insert(attr.oid.clone(), attr).is_some() {
+            // duplicate attributes are not allowed
+            return Err(Err::Failure(X509Error::DuplicateAttributes));
+        }
+    }
+    Ok((i, attributes))
+}
+
+fn parse_cri_attributes(i: &[u8]) -> X509Result<HashMap<Oid, X509CriAttribute>> {
+    let (i, hdr) = ber_read_element_header(i).or(Err(Err::Error(X509Error::InvalidAttributes)))?;
+    (0..hdr.structured)
+        .into_iter()
```

Comment:
If I get that correctly, this is equivalent to a `if hdr.structured != 0` statement ?

Reply:
I guess it might be, I probably don't have the same level of ASN.1 understanding you do. :)

Reply:
Also note that I have only tested this on CSRs that have exactly one attribute, so this could perhaps benefit from testing with 0 and more-than-1 attributes (though I'm not clear if any other attributes are commonly in use).